fp.go

```diff
@@ -0,0 +1,411 @@
+	}
+
+	x.MulUnsafe(a, b)
+}
+
+// MulUnsafe computes the product of the two field elements and stores the
+// result in the receiver.
+//
+// Unsafe: If this function receives nil arguments, the behaviour is
+// implementation dependent, because the definition of the NULL pointer in c is
+// implementation dependent. If the NULL pointer and the go nil pointer are the
+// same, then the function will panic.
+func (x *Fp) MulUnsafe(a, b *Fp) {
+	// NOTE: The c function defines the pointer to b as restrict, which means
+	// that it must not be aliased by x or a. We therefore make a new copy of b
+	// to ensure that there is no aliasing.
+	bCopy := *b
+	C.secp256k1_fe_mul(
+		(*C.secp256k1_fe)(unsafe.Pointer(x)),
+		(*C.secp256k1_fe)(unsafe.Pointer(a)),
+		(*C.secp256k1_fe)(unsafe.Pointer(&bCopy)),
+	)
+	x.normalize()
+}
+
+// Sqr computes the square of the field element and stores the result in the
+// receiver.
+func (x *Fp) Sqr(a *Fp) {
+	if a == nil {
+		panic("expected first argument to be not be nil")
+	}
+
+	x.SqrUnsafe(a)
+}
+
+// SqrUnsafe computes the square of the field element and stores the result in
+// the receiver.
+//
+// Unsafe: If this function receives nil arguments, the behaviour is
+// implementation dependent, because the definition of the NULL pointer in c is
+// implementation dependent. If the NULL pointer and the go nil pointer are the
+// same, then the function will panic.
+func (x *Fp) SqrUnsafe(a *Fp) {
+	C.secp256k1_fe_sqr((*C.secp256k1_fe)(unsafe.Pointer(x)), (*C.secp256k1_fe)(unsafe.Pointer(a)))
+	x.normalize()
+}
+
+// Inv computes the multiplicative inverse of the field element and stores the
+// result in the receiver.
+func (x *Fp) Inv(a *Fp) {
+	if a == nil {
+		panic("expected first argument to be not be nil")
+	}
+
+	x.InvUnsafe(a)
+}
+
+// InvUnsafe computes the multiplicative inverse of the field element and
+// stores the result in the receiver.
+//
+// Unsafe: If this function receives nil arguments, the behaviour is
+// implementation dependent, because the definition of the NULL pointer in c is
+// implementation dependent. If the NULL pointer and the go nil pointer are the
+// same, then the function will panic.
+func (x *Fp) InvUnsafe(a *Fp) {
+	// We use the potentially faster but not constant time version of the
+	// inverse function.
+	C.secp256k1_fe_inv_var((*C.secp256k1_fe)(unsafe.Pointer(x)), (*C.secp256k1_fe)(unsafe.Pointer(a)))
+	x.normalize()
+}
+
+// IsZero returns true if the field element is zero and false otherwise.
+func (x *Fp) IsZero() bool {
+	return (x.inner[0] | x.inner[1] | x.inner[2] | x.inner[3] | x.inner[4]) == 0
+}
+
+// IsOne returns true if the field element is zero and false otherwise.
+func (x *Fp) IsOne() bool {
+	return (x.inner[0] == 1) && ((x.inner[1] | x.inner[2] | x.inner[3] | x.inner[4]) == 0)
+}
+
+// IsEven returns true if the field element is even and false otherwise.
+func (x *Fp) IsEven() bool {
+	return !((x.inner[0] & 1) == 1)
```

Comment:
Confirming, this is meant to be `x.inner[0]` and not `x.inner[4]`?

Reply:
Yep, `x.inner[0]` is the least significant limb.